File: go/solutions/2130-maximum-twin-sum-of-a-linked-list.go
package solutions

import (
	"github.com/SamuelMuloki/LeetCode/go/utils"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func PairSum(head *utils.ListNode) int {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	second, prev := slow, new(utils.ListNode)
	for second != nil {
		tmp := second.Next
		second.Next = prev
		prev = second
		second = tmp
	}

	maxSum := 0
	for head != slow && prev != nil {
		maxSum = max(maxSum, head.Val+prev.Val)
		head = head.Next
		prev = prev.Next
	}

	return maxSum
}

File: js/2130-maximum-twin-sum-of-a-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function (head) {
  var slow = (fast = head);
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  var second = slow,
    prev = null;
  while (second) {
    var tmp = second.next;
    second.next = prev;
    prev = second;
    second = tmp;
  }

  var maxSum = 0;
  while (prev) {
    maxSum = Math.max(maxSum, head.val + prev.val);
    head = head.next;
    prev = prev.next;
  }

  return maxSum;
};

module.exports = { pairSum };
